namerd/storage/k8s/src/main/scala/io/buoyant/namerd/storage/experimental/K8sDtabStore.scala

```diff
@@ -0,0 +1,136 @@
+ *
+ * This creates a third-party resource at: /apis/buoyant.io/v1/dtabs.
+ *
+ * TODO: add tests: unit, end-to-end
+ *
+ * @param client    a fully configured client which will be used to access the Kubernetes API
+ * @param namespace the Kubernetes namespace being read/written by this store (note that this differs from Dtab namespaces)
+ */
+class K8sDtabStore(client: Http.Client, dst: String, namespace: String)
+  extends DtabStore {
+  private[this] val log = Logger()
+
+  // NOTE: we construct two clients, one for the streaming watch requests, one for everything else.
+  // This is due to an as-yet-undetermined issue where the first non-watch request will hang if using
+  // the same client.
+  val api: NsApi = Api(client.newService(dst)).namespace("default")
+  val watchApi: NsApi = Api(client.newService(dst)).namespace("default")
+
+  private[this] def name(dtab: Dtab): String = {
+    dtab.metadata.flatMap(_.name).getOrElse {
+      throw new IllegalStateException("k8s API returned a dtab with no name: $dtab")
+    }
+  }
+
+  private[this] def toDtabMap(dtab: Dtab): (String, VersionedDtab) = {
+    val vsn = dtab.metadata.flatMap(_.resourceVersion).getOrElse {
+      throw new IllegalStateException("k8s API returned a resource with no version")
+    }
+    val versionedDtab = VersionedDtab(FDtab(dtab.dentries.toIndexedSeq), Buf.Utf8(vsn))
+    (name(dtab), versionedDtab)
+  }
+
+  private[this]type NsMap = Map[String, VersionedDtab]
+  private[this] val (act: Activity[NsMap], witness: Witness[Try[NsMap]]) = Activity[NsMap]()
+
+  // NOTE: this currently relies on our watch connection to update the internal state, rather than doing
+  // any kind of "write-through caching", so there are currently no "read-your-write" semantics. If those
+  // are required, we would probably need to implement a synchronized means of reading `act` and using that value to
+  // feed to `witness`, rather than taking the `foldLeft` approach as below.
+  api.dtabs.get(None, None, None, retryIndefinitely = true).onSuccess { dtabList =>
+    val initState: NsMap = dtabList.items.map(toDtabMap)(breakOut)
+    witness.notify(Return(initState))
+    val (stream, close) = watchApi.dtabs.watch(None, None, dtabList.apiVersion)
+    stream.foldLeft(initState) { (nsMap, watchEvent) =>
+      val newState: NsMap = watchEvent match {
+        case DtabWatch.Added(a) => nsMap + toDtabMap(a)
+        case DtabWatch.Modified(m) => nsMap + toDtabMap(m)
+        case DtabWatch.Deleted(d) => nsMap - name(d)
+        case DtabWatch.Error(e) =>
+          log.error("k8s watch error: %s", e)
+          nsMap
+      }
+      witness.notify(Return(newState))
+      newState
+    }
+  }
+
+  /** List all Dtab namespaces */
+  def list(): Future[Set[String]] = act.run.sample match {
+    case Activity.Ok(nses) => Future.value(nses.keySet)
+    case Activity.Pending => for {
+      tryNses: Try[NsMap] <- act.values.toFuture
+      nses <- Future.const(tryNses)
+    } yield nses.keySet
```

Comment:
nit: would be slightly easier to read if the whole `for { }` loop was indented:

e.g
```scala
case Activity.Pending =>
  for {
    ..
  } yield ..
case ...
```

somewhat visually confusing that the yield is at the same indent level as the next case